### ISSUE TITLE
Performance issue with MOE models

### ISSUE THREAD

**Issue by teux91** [JamePeng/llama-cpp-python#123]
Hi

with a rtx 5060ti

tried some text only prompt with gemma4-26B-A4B MXFP4 from unsloth using lm studio and officaial llama.ccp cli get around 45 t/s with  all layers in gpu and all moe layers on cpu 

if using latest cuda build of llama-cpp-python initialized like this:

llm= Llama(
                    model_path="./gemma-4-26B-A4B-it-MXFP4_MOE.gguf",
                    n_ctx=4096,
                    n_threads=6,
                    n_gpu_layers=-1,
                    n_cpu_moe=30,
                    flash_attn_type=1,
                    logits_all=False,
                    verbose=True,
                    use_mmap=False,
                )

got only around 10 t/S during generation

same thing with qwen3.6-35B, same thing with other quant type

Does anyone have same problem, am i doing something wrong ?

**Comment by JamePeng** [comment]
Please provide the specific system and hardware configuration (including CPU and RAM), the video memory and system memory usage during operation, and screenshots of processor usage.

**Comment by JamePeng** [comment]
You can try version 0.3.38 again for comparison.

I also recommend compiling and installing it locally, as the default basic version doesn't include any AVX instruction set enhancements for compatibility reasons, which may result in performance loss.

**Comment by deepfree2023** [comment]
Hi, when will there be an AVX2 version?

Pre-built Wheel (New)

It is also possible to install a pre-built wheel with CUDA support. As long as your system meets some requirements:

CUDA Version is 12.4, 12.6, 12.8 or 13.0
Python Version is 3.10, 3.11, 3.12, 3.13 or 3.14
Basic version(Default): A version compiled without using AVX instructions (for compatibility with CPU platforms lacking AVX instructions or with AVX instruction compatibility issues).
AVX2 version: A version compiled using AVX2 instructions.

**Comment by teux91** [comment]
Hi,

thank you for your reply, still same problem with  0.3.38
proc is AMD Ryzen 5 7500F 6-Core 

llama.cpp used in lm studio:
"name": "llama.cpp-win-x86_64-nvidia-cuda12-avx2",
  "version": "2.13.0",
  "platform": "win",
  "cpu": {
    "architecture": "x86_64",
    "instruction_set_extensions": [
      "AVX2"
    ]
  },


enclosed screenshots of lmstudio config ram vram cpu gpu usage and resultant speed of 43t/s

for llama-cpp-python tried too options
with option 1 same config as lm studio all layers on gpu and all moe layers on cpu, vram and gpu usage too low  4,3 t/s

tried option 2 with all layers to gpu and all 10 moe layers on cpu , vram usafe increase to 14,7GB but GPU usage is only 25% compared to 45% with lmstrudio and speed at 15t/sec only


could it be because of the lack of avx2 in llama-cpp-python? why llama-cpp-python option 1 config do not give same vram usage as lm studio ?

<img width="1951" height="1088" alt="Image" src="https://github.com/user-attachments/assets/b76656fb-1087-4259-abe4-4dc511f1096e" />
<img width="1887" height="1043" alt="Image" src="https://github.com/user-attachments/assets/164807b2-74a0-4c0b-83ff-ab2c26928843" />
<img width="1630" height="265" alt="Image" src="https://github.com/user-attachments/assets/99bcda7e-42b9-40c6-a981-da05ed276b54" />
<img width="937" height="1125" alt="Image" src="https://github.com/user-attachments/assets/647606bd-7397-48cd-b903-aa7f32884ebe" />
<img width="1736" height="1032" alt="Image" src="https://github.com/user-attachments/assets/474fc3de-886d-43e1-a56b-f168b14b42e8" />
<img width="1516" height="847" alt="Image" src="https://github.com/user-attachments/assets/3e685306-705b-4ccd-a58f-2ee59f2c0487" />

**Comment by teux91** [comment]
ok my bad recompiled with cuda and avx512 support got 60t/sec

**Comment by JamePeng** [comment]
@deepfree2023 I had compiled it before, but the compilation based on the GitHub virtual machine does not represent the user's local CPU environment. It often causes crashes due to poor compatibility with AMD's AVX, so I still recommend compiling locally.

**Comment by JamePeng** [comment]
@deepfree2023 @teux91 Another method you can try, which requires a little manual work, is to use [ggml-org/llama.cpp](https://github.com/ggml-org/llama.cpp/releases) to select the CUDA version you need from the compressed package, and copy all the DLLs inside to your path \Python\Python3XX\Lib\site-packages\llama_cpp\lib. For conda or comfyui, you need to find the installation path and replace them.

<img width="1236" height="1239" alt="Image" src="https://github.com/user-attachments/assets/b0a9a1b6-878d-43bc-b31d-9ff7c1aa7b0b" />

**Comment by teux91** [comment]
Hello Jame,

all fine now with locally compiled version, thanks a lot for your great work and your fast and positive replies
